File: CS/WebSite/App_Code/DataHelper.cs
using System;
using System.Data.OleDb;
using System.Web.UI;
using System.Web.UI.WebControls;
using DevExpress.XtraScheduler;
using DevExpress.Web.ASPxScheduler;
using DevExpress.Web.ASPxScheduler.Internal;
using System.Configuration;


#region DataHelper
public static class DataHelper {
	public static void SetupDefaultMappings(ASPxScheduler control) {
		SetupDefaultMappingsSiteMode(control);
	}
	private static void SetupDefaultMappingsSiteMode(ASPxScheduler control) {
		ASPxSchedulerStorage storage = control.Storage;
		storage.BeginUpdate();
		try {
			ASPxResourceMappingInfo resourceMappings = storage.Resources.Mappings;
			resourceMappings.ResourceId = "ID";
			resourceMappings.Caption = "Model";

			ASPxAppointmentMappingInfo appointmentMappings = storage.Appointments.Mappings;
			appointmentMappings.AppointmentId = "ID";
			appointmentMappings.Start = "StartTime";
			appointmentMappings.End = "EndTime";
			appointmentMappings.Subject = "Subject";
			appointmentMappings.AllDay = "AllDay";
			appointmentMappings.Description = "Description";
			appointmentMappings.Label = "Label";
			appointmentMappings.Location = "Location";
			appointmentMappings.RecurrenceInfo = "RecurrenceInfo";
			appointmentMappings.ReminderInfo = "ReminderInfo";
			appointmentMappings.ResourceId = "CarId";
			appointmentMappings.Status = "Status";
			appointmentMappings.Type = "EventType";
		}
		finally {
			storage.EndUpdate();
		}
	}
	public static void ProvideRowInsertion(ASPxScheduler control, DataSourceControl dataSource) {
		ObjectDataSource objectDataSource = dataSource as ObjectDataSource;
		if (objectDataSource != null) {
			ObjectDataSourceRowInsertionProvider provider = new ObjectDataSourceRowInsertionProvider();
			provider.ProvideRowInsertion(control, objectDataSource);
		}
	}
}
#endregion
#region ObjectDataSourceRowInsertionProvider
public class ObjectDataSourceRowInsertionProvider {
	int lastInsertedAppointmentId;
	public void ProvideRowInsertion(ASPxScheduler control, ObjectDataSource dataSource) {
		dataSource.Inserted += new ObjectDataSourceStatusEventHandler(AppointmentsDataSource_Inserted);
		control.AppointmentRowInserted += new ASPxSchedulerDataInsertedEventHandler(ControlOnAppointmentRowInserted);
		control.AppointmentsInserted += new PersistentObjectsEventHandler(ControlOnAppointmentsInserted);
	}
	void ControlOnAppointmentRowInserted(object sender, ASPxSchedulerDataInsertedEventArgs e) {
		// Autoincremented primary key case
		e.KeyFieldValue = this.lastInsertedAppointmentId;
	}
	void AppointmentsDataSource_Inserted(object sender, ObjectDataSourceStatusEventArgs e) {
		// Autoincremented primary key case
		this.lastInsertedAppointmentId = (int)e.ReturnValue;
	}
	void ControlOnAppointmentsInserted(object sender, PersistentObjectsEventArgs e) {
		//Autoincremented primary key case
		int count = e.Objects.Count;
		System.Diagnostics.Debug.Assert(count == 1);
		Appointment apt = (Appointment)e.Objects[0];
		ASPxSchedulerStorage storage = (ASPxSchedulerStorage)sender;
		storage.SetAppointmentId(apt, lastInsertedAppointmentId);
	}
}
#endregion
#region DbAdapterHelperBase
public class DbAdapterHelperBase {
	#region Fields
	OleDbDataAdapter adapter;
	OleDbConnection connection;
	#endregion

	protected DbAdapterHelperBase() {
		this.adapter = new OleDbDataAdapter();
		this.connection = CreateConnection();

		InitAdapter(Adapter, Connection);
	}

	#region Properties
	public OleDbDataAdapter Adapter { get { return adapter; } }
	public OleDbConnection Connection { get { return connection; } }
	#endregion

	#region CreateConnection
	protected OleDbConnection CreateConnection() {
		OleDbConnection connection = new OleDbConnection();
		connection.ConnectionString = ConfigurationManager.ConnectionStrings["SchedulerConnectionString"].ConnectionString; 
		return connection;
	}
	#endregion
	#region InitAdapter
	protected void InitAdapter(OleDbDataAdapter adapter, OleDbConnection connection) {
		adapter.SelectCommand = CreateSelectionCommand(connection);
		adapter.DeleteCommand = CreateDeleteCommand(connection);
		adapter.UpdateCommand = CreateUpdateCommand(connection);
		adapter.InsertCommand = CreateInsertCommand(connection);
	}
	#endregion
	#region CreateSelectionCommand
	protected virtual OleDbCommand CreateSelectionCommand(OleDbConnection connection) {
		return null;
	}
	#endregion
	#region CreateDeleteCommand
	protected virtual OleDbCommand CreateDeleteCommand(OleDbConnection connection) {
		return null;
	}
	#endregion
	#region CreateUpdateCommand
	protected virtual OleDbCommand CreateUpdateCommand(OleDbConnection connection) {
		return null;
	}
	#endregion
	#region CreateInsertCommand
	protected virtual OleDbCommand CreateInsertCommand(OleDbConnection connection) {
		return null;
	}
	#endregion
}
#endregion

File: VB/WebSite/App_Code/DataHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Data.OleDb
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports DevExpress.XtraScheduler
Imports DevExpress.Web.ASPxScheduler
Imports DevExpress.Web.ASPxScheduler.Internal
Imports System.Configuration


#Region "DataHelper"
Public NotInheritable Class DataHelper
	Private Sub New()
	End Sub
	Public Shared Sub SetupDefaultMappings(ByVal control As ASPxScheduler)
		SetupDefaultMappingsSiteMode(control)
	End Sub
	Private Shared Sub SetupDefaultMappingsSiteMode(ByVal control As ASPxScheduler)
		Dim storage As ASPxSchedulerStorage = control.Storage
		storage.BeginUpdate()
		Try
			Dim resourceMappings As ASPxResourceMappingInfo = storage.Resources.Mappings
			resourceMappings.ResourceId = "ID"
			resourceMappings.Caption = "Model"

			Dim appointmentMappings As ASPxAppointmentMappingInfo = storage.Appointments.Mappings
			appointmentMappings.AppointmentId = "ID"
			appointmentMappings.Start = "StartTime"
			appointmentMappings.End = "EndTime"
			appointmentMappings.Subject = "Subject"
			appointmentMappings.AllDay = "AllDay"
			appointmentMappings.Description = "Description"
			appointmentMappings.Label = "Label"
			appointmentMappings.Location = "Location"
			appointmentMappings.RecurrenceInfo = "RecurrenceInfo"
			appointmentMappings.ReminderInfo = "ReminderInfo"
			appointmentMappings.ResourceId = "CarId"
			appointmentMappings.Status = "Status"
			appointmentMappings.Type = "EventType"
		Finally
			storage.EndUpdate()
		End Try
	End Sub
	Public Shared Sub ProvideRowInsertion(ByVal control As ASPxScheduler, ByVal dataSource As DataSourceControl)
		Dim objectDataSource As ObjectDataSource = TryCast(dataSource, ObjectDataSource)
		If objectDataSource IsNot Nothing Then
			Dim provider As New ObjectDataSourceRowInsertionProvider()
			provider.ProvideRowInsertion(control, objectDataSource)
		End If
	End Sub
End Class
#End Region
#Region "ObjectDataSourceRowInsertionProvider"
Public Class ObjectDataSourceRowInsertionProvider
	Private lastInsertedAppointmentId As Integer
	Public Sub ProvideRowInsertion(ByVal control As ASPxScheduler, ByVal dataSource As ObjectDataSource)
		AddHandler dataSource.Inserted, AddressOf AppointmentsDataSource_Inserted
		AddHandler control.AppointmentRowInserted, AddressOf ControlOnAppointmentRowInserted
		AddHandler control.AppointmentsInserted, AddressOf ControlOnAppointmentsInserted
	End Sub
	Private Sub ControlOnAppointmentRowInserted(ByVal sender As Object, ByVal e As ASPxSchedulerDataInsertedEventArgs)
		' Autoincremented primary key case
		e.KeyFieldValue = Me.lastInsertedAppointmentId
	End Sub
	Private Sub AppointmentsDataSource_Inserted(ByVal sender As Object, ByVal e As ObjectDataSourceStatusEventArgs)
		' Autoincremented primary key case
		Me.lastInsertedAppointmentId = CInt(Fix(e.ReturnValue))
	End Sub
	Private Sub ControlOnAppointmentsInserted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
		'Autoincremented primary key case
		Dim count As Integer = e.Objects.Count
		System.Diagnostics.Debug.Assert(count = 1)
		Dim apt As Appointment = CType(e.Objects(0), Appointment)
		Dim storage As ASPxSchedulerStorage = CType(sender, ASPxSchedulerStorage)
		storage.SetAppointmentId(apt, lastInsertedAppointmentId)
	End Sub
End Class
#End Region
#Region "DbAdapterHelperBase"
Public Class DbAdapterHelperBase
	#Region "Fields"
	Private adapter_Renamed As OleDbDataAdapter
	Private connection_Renamed As OleDbConnection
	#End Region

	Protected Sub New()
		Me.adapter_Renamed = New OleDbDataAdapter()
		Me.connection_Renamed = CreateConnection()

		InitAdapter(Adapter, Connection)
	End Sub

	#Region "Properties"
	Public ReadOnly Property Adapter() As OleDbDataAdapter
		Get
			Return adapter_Renamed
		End Get
	End Property
	Public ReadOnly Property Connection() As OleDbConnection
		Get
			Return connection_Renamed
		End Get
	End Property
	#End Region

	#Region "CreateConnection"
	Protected Function CreateConnection() As OleDbConnection
		Dim connection As New OleDbConnection()
		connection.ConnectionString = ConfigurationManager.ConnectionStrings("SchedulerConnectionString").ConnectionString
		Return connection
	End Function
	#End Region
	#Region "InitAdapter"
	Protected Sub InitAdapter(ByVal adapter As OleDbDataAdapter, ByVal connection As OleDbConnection)
		adapter.SelectCommand = CreateSelectionCommand(connection)
		adapter.DeleteCommand = CreateDeleteCommand(connection)
		adapter.UpdateCommand = CreateUpdateCommand(connection)
		adapter.InsertCommand = CreateInsertCommand(connection)
	End Sub
	#End Region
	#Region "CreateSelectionCommand"
	Protected Overridable Function CreateSelectionCommand(ByVal connection As OleDbConnection) As OleDbCommand
		Return Nothing
	End Function
	#End Region
	#Region "CreateDeleteCommand"
	Protected Overridable Function CreateDeleteCommand(ByVal connection As OleDbConnection) As OleDbCommand
		Return Nothing
	End Function
	#End Region
	#Region "CreateUpdateCommand"
	Protected Overridable Function CreateUpdateCommand(ByVal connection As OleDbConnection) As OleDbCommand
		Return Nothing
	End Function
	#End Region
	#Region "CreateInsertCommand"
	Protected Overridable Function CreateInsertCommand(ByVal connection As OleDbConnection) As OleDbCommand
		Return Nothing
	End Function
	#End Region
End Class
#End Region

